### PR TITLE
fix: Implemented Rate Limiter against Message Flooding.

### LINKED ISSUE #470: Implement Rate Limiting for Messages.
### Description

Without rate limiting, users can flood the chat with messages, potentially leading to performance issues and misuse. Rate limiting will ensure that users can send messages at a reasonable pace and prevent abuse.

### Screenshots

anon_chat

### Additional information

I propose implementing rate limiting for message sending. A rate limit of, for example, 70 messages per minute per user, can be a good starting point .

### 🥦 Browser

Google Chrome

### 👀 Have you checked if this issue has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Contributing Guidelines?

- [X] I have read and understood the rules in the [Contributing Guidelines](https://github.com/Dun-sin/Whisper/blob/main/CONTRIBUTING.md)

### Are you willing to work on this issue ?

Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by Tholkappiar**
# Fixes Issue

**My PR closes #470 **

# 👨‍💻 Changes proposed(What did you do ?)

# ✔️ Check List (Check all the applicable boxes)

- [x] My code follows the code style of this project.
- [x] This PR does not contain plagiarized content.
- [x] The title and description of the PR is clear and explains the approach.

##  Note to reviewers

The message will not be sent when it hit the count 25 in a minute.

# 📷 Screenshots

WhatsApp Image 2023-10-20 at 7 54 04 AM

### CODE REVIEW

**Comment by Tholkappiar** [comment]
Hello, Maintainers!

I hope you are doing well. I've submitted a pull request by Adding the Rate Limiter for Messages .Please feel free to review my changes and provide any feedback you may have.

Thank you for your time and consideration. I look forward to hearing from you soon.

Best regards,
Tholkappiar

**Comment by Dun-sin** [comment]
> Hello, Maintainers!
> 
> I hope you are doing well. I've submitted a pull request by Adding the Rate Limiter for Messages .Please feel free to review my changes and provide any feedback you may have.
> 
> Thank you for your time and consideration. I look forward to hearing from you soon.
> 
> _PS : Kindly Add up Hacktoberfest things for this PR !_
> 
> Best regards, Tholkappiar

just talk like a normal person, no need for the email

File: server/package.json [hunk 1]
```diff
@@ -31,9 +32,9 @@
   "devDependencies": {
     "eslint": "^8.47.0",
     "eslint-config-prettier": "^9.0.0",
+    "husky": "^8.0.3",
     "lint-staged": "^14.0.1",
-    "prettier": "^3.0.1",
-    "husky": "^8.0.3"
+    "prettier": "^3.0.1"
```
Comment (Dun-sin) [inline server/package.json:37]: why the change here?

File: server/sockets/sendMessage.js [hunk 2]
```diff
@@ -1,56 +1,61 @@
-const {
-  NEW_EVENT_SEND_MESSAGE,
-  NEW_EVENT_SEND_FAILED,
-  NEW_EVENT_RECEIVE_MESSAGE,
-} = require('../../constants.json');
-const { addMessage, getActiveUser } = require('../utils/lib');
+const { NEW_EVENT_SEND_MESSAGE, NEW_EVENT_SEND_FAILED, NEW_EVENT_RECEIVE_MESSAGE } = require("../../constants.json");
```
Comment (Dun-sin) [inline server/sockets/sendMessage.js:1]: use the proper formatting, do a git pull, install prettier and update the result

File: server/sockets/sendMessage.js [hunk 3]
```diff
@@ -1,56 +1,61 @@
...
+      }
 
+      // Check the message count for the user
+      const userMessageCount = messageCounts[senderId] || 0;
+
+      if (userMessageCount >= 25) {
+        // User has exceeded the message limit
+        socket.emit(NEW_EVENT_SEND_FAILED, {
+          message: "You have exceeded the message limit. Please try again later.",
+        });
```
Comment (Dun-sin) [inline server/sockets/sendMessage.js:30]: you also need to give the alert to the user in the frontend

**Review by Dun-sin — CHANGES_REQUESTED** [review]

File: server/sockets/sendMessage.js (hunk 3 above)
Comment (Tholkappiar) [reply server/sockets/sendMessage.js:30]: Okay...👍🏻

File: server/package.json (hunk 1 above)
Comment (Tholkappiar) [reply server/package.json:37]: "express-rate-limit": "^7.1.1" - for the rate limit ! any other things kindly reply?

File: server/sockets/sendMessage.js (hunk 2 above)
Comment (Tholkappiar) [reply server/sockets/sendMessage.js:1]: okie , I will update the code format with prettier !

File: server/sockets/sendMessage.js (hunk 3 above)
Comment (Tholkappiar) [reply server/sockets/sendMessage.js:30]: What kind of alert messages ! normal alert or using react-toastify ?

Comment (Dun-sin) [reply server/sockets/sendMessage.js:30]: normal alert

Comment (Dun-sin) [reply server/sockets/sendMessage.js:30]: remember to do a git pull before a git push

File: server/package.json (hunk 1 above)
Comment (Dun-sin) [reply server/package.json]: this is what the screenshot says, your answer isn't related to the question:
image
